modules/gdscript/language_server/gdscript_workspace.cpp

```diff
@@ -458,48 +436,108 @@ Error GDScriptWorkspace::parse_script(const String &p_path, const String &p_cont
-	if (reference_symbol) {
-		String identifier = reference_symbol->name;
+	if (is_valid_rename_target(reference_symbol)) {
+		Vector<lsp::Location> usages = find_all_usages(*reference_symbol);
+		for (int i = 0; i < usages.size(); ++i) {
+			lsp::Location loc = usages[i];
 
-		for (List<String>::Element *PE = paths.front(); PE; PE = PE->next()) {
-			PackedStringArray content = FileAccess::get_file_as_string(PE->get(), &err).split("\n");
-			for (int i = 0; i < content.size(); ++i) {
-				String line = content[i];
+			edit.add_change(loc.uri, loc.range.start.line, loc.range.start.character, loc.range.end.character, new_name);
+		}
+	}
 
-				int character = line.find(identifier);
-				while (character > -1) {
-					lsp::TextDocumentPositionParams params;
+	return edit.to_json();
+}
 
-					lsp::TextDocumentIdentifier text_doc;
-					text_doc.uri = get_file_uri(PE->get());
+bool GDScriptWorkspace::can_rename(const lsp::TextDocumentPositionParams &p_doc_pos, lsp::DocumentSymbol &r_symbol, lsp::Range &r_range) {
+	const lsp::DocumentSymbol *reference_symbol = resolve_symbol(p_doc_pos);
+	if (!is_valid_rename_target(reference_symbol)) {
+		return false;
+	}
 
-					params.textDocument = text_doc;
-					params.position.line = i;
-					params.position.character = character;
+	String path = get_file_path(p_doc_pos.textDocument.uri);
+	if (const ExtendGDScriptParser *parser = get_parse_result(path)) {
+		parser->get_identifier_under_position(p_doc_pos.position, r_range);
+		r_symbol = *reference_symbol;
+		return true;
+	}
 
-					const lsp::DocumentSymbol *other_symbol = resolve_symbol(params);
+	return false;
+}
 
-					if (other_symbol == reference_symbol) {
-						edit.add_change(text_doc.uri, i, character, character + identifier.length(), new_name);
-					}
+Vector<lsp::Location> GDScriptWorkspace::find_usages_in_file(const lsp::DocumentSymbol &p_symbol, const String &p_file_path) {
+	Vector<lsp::Location> usages;
 
-					character = line.find(identifier, character + 1);
-				}
+	String identifier = p_symbol.name;
+	Error err;
+	PackedStringArray content = FileAccess::get_file_as_string(p_file_path, &err).split("\n");
+	for (int i = 0; i < content.size(); ++i) {
+		String line = content[i];
+
+		int character = line.find(identifier);
+		while (character > -1) {
+			lsp::TextDocumentPositionParams params;
+
+			lsp::TextDocumentIdentifier text_doc;
+			text_doc.uri = get_file_uri(p_file_path);
+
+			params.textDocument = text_doc;
+			params.position.line = i;
+			params.position.character = character;
+
+			const lsp::DocumentSymbol *other_symbol = resolve_symbol(params);
+
+			if (other_symbol == &p_symbol) {
+				lsp::Location loc;
+				loc.uri = text_doc.uri;
+				loc.range.start = params.position;
+				loc.range.end.line = params.position.line;
+				loc.range.end.character = params.position.character + identifier.length();
+				usages.append(loc);
 			}
+
+			character = line.find(identifier, character + 1);
 		}
 	}
```

Comment:
This here is the issue with renames: It reads the source from disk... Of course it then only  works once a change is saved.... 😳 

Should get cached lines instead (-> inside parser)

-> replace `content = FileAccess::get_file_as_string(...` with
```cpp
	if (const ExtendGDScriptParser *parser = get_parse_result(p_file_path)) {
		const PackedStringArray &content = parser->get_lines();
```


```suggestion
	if (const ExtendGDScriptParser *parser = get_parse_result(p_file_path)) {
		const PackedStringArray &content = parser->get_lines();
		for (int i = 0; i < content.size(); ++i) {
			String line = content[i];

			int character = line.find(identifier);
			while (character > -1) {
				lsp::TextDocumentPositionParams params;

				lsp::TextDocumentIdentifier text_doc;
				text_doc.uri = get_file_uri(p_file_path);

				params.textDocument = text_doc;
				params.position.line = i;
				params.position.character = character;

				const lsp::DocumentSymbol *other_symbol = resolve_symbol(params);

				if (other_symbol == &p_symbol) {
					lsp::Location loc;
					loc.uri = text_doc.uri;
					loc.range.start = params.position;
					loc.range.end.line = params.position.line;
					loc.range.end.character = params.position.character + identifier.length();
					usages.append(loc);
				}

				character = line.find(identifier, character + 1);
			}
		}
	}
```